src/main/java/macrobase/analysis/MacrobaseAnalyzer.java

```diff
@@ -0,0 +1,55 @@
+package macrobase.analysis;
+
+import macrobase.analysis.outlier.OutlierClassifier;
+import macrobase.analysis.result.MacrobaseAnalysisResult;
+import macrobase.analysis.summary.Summarizer;
+import macrobase.conf.MacroBaseConf;
+import macrobase.datamodel.Datum;
+import macrobase.ingest.DataIngester;
+import macrobase.transform.FeatureTransform;
+import org.apache.commons.collections4.iterators.IteratorIterable;
+
+import java.util.Iterator;
+
+public class MacrobaseAnalyzer implements Iterator<MacrobaseAnalysisResult> {
+
+    private DataIngester ingester;
+    private FeatureTransform source;
+    private OutlierClassifier outlierClassifier;
+    private Summarizer summarizer;
+    private int windowSize;
+
+    /**
+     * MacrobaseAnalyzer will create the following attributes in the constructor:
+     *  - DataIngester ingester, that gets the data, usually just initialize with a conf, based on conf
+     *  - FeatureTransform based on conf and ingester
+     * @param conf
+     */
+    public MacrobaseAnalyzer(MacroBaseConf conf) {};
+    // Example pseudo-code,
+    // ingester = constructIngester(conf);
+    // source = new FeatureTransform(conf, ingester);
+    // windowSize = conf.getWindowSize();
+    // outlierClassifier = constructOutlierClassifier(conf);
+    // Summarizer = new Summarizer(conf, outlierClassifier);
+
+    @Override
+    public MacrobaseAnalysisResult next() {
+
+        int index = 0;
+        for (Datum nextDatum : new IteratorIterable<>(source)) {
+            summarizer.observe(nextDatum);
+            if (index > windowSize) {
```

Comment:
Actually, ignore the incremental summarization. To be clearer, for this and my comment on the above line, I was thinking of making `Summarizer` an iterator. So like most of the other classes, you could then call next on it and it would automatically fetch the next input from the classifier so that this explicit call to `observe` isn't required.

So the new pipeline is Ingest -> Feature extraction -> Context selection -> Density estimation -> Summarization. Once you've constructed the pipeline, the idea is just to call `next` on the last step of the pipeline, and each step calls `next` on the previous step as required. I think that makes for a clean and modular architecture, both conceptually and code-wise.